packages/service-library/src/servicelib/long_running_tasks/task.py

```diff
@@ -68,75 +72,126 @@ async def _await_task(task: asyncio.Task) -> None:
     await task
 
 
-def _get_tasks_to_remove(
-    tracked_tasks: TrackedTaskGroupDict,
+async def _get_tasks_to_remove(
+    tracked_tasks: BaseStore,
     stale_task_detect_timeout_s: PositiveFloat,
-) -> list[TaskId]:
+) -> list[tuple[TaskId, TaskContext]]:
     utc_now = datetime.datetime.now(tz=datetime.UTC)
 
-    tasks_to_remove: list[TaskId] = []
+    tasks_to_remove: list[tuple[TaskId, TaskContext]] = []
 
-    for task_id, tracked_task in tracked_tasks.items():
+    for tracked_task in await tracked_tasks.list_tasks_data():
         if tracked_task.fire_and_forget:
             continue
 
         if tracked_task.last_status_check is None:
             # the task just added or never received a poll request
             elapsed_from_start = (utc_now - tracked_task.started).seconds
             if elapsed_from_start > stale_task_detect_timeout_s:
-                tasks_to_remove.append(task_id)
+                tasks_to_remove.append(
+                    (tracked_task.task_id, tracked_task.task_context)
+                )
         else:
             # the task status was already queried by the client
             elapsed_from_last_poll = (utc_now - tracked_task.last_status_check).seconds
             if elapsed_from_last_poll > stale_task_detect_timeout_s:
-                tasks_to_remove.append(task_id)
+                tasks_to_remove.append(
+                    (tracked_task.task_id, tracked_task.task_context)
+                )
     return tasks_to_remove
 
 
-class TasksManager:
+def _error_to_string(e: Exception) -> str:
+    return base64.b85encode(pickle.dumps(e)).decode("utf-8")
+
+
+def _error_from_string(error_str: str) -> Exception:
+    return pickle.loads(base64.b85decode(error_str))  # type: ignore[no-any-return]  # noqa: S301
+
+
+class TasksManager:  # pylint:disable=too-many-instance-attributes
```

Comment:
Refactoring TIP: The core idea is that each worker task is independent but follows a common lifecycle pattern (e.g., setup, teardown, event signaling). I recommend encapsulating this shared behavior into a dedicated class, allowing each worker to manage its own lifecycle. These worker instances can then be managed collectively by a TasksManager. The current flat implementation mixes responsibilities and makes the code harder to scale and maintain.

Reply:
I'm really unsure why this flat implementation makes it harder. I find what you propose harder to achieve. I agree it was a bit messy and I've tried to clean it up a bit.